### PR TITLE
Add usage statuses endpoint #230

### LINKED ISSUE #230: Add usage statuses endpoint
see #170

### PR DESCRIPTION
**Pull request by joshuadkitenge**
## Description
- Add collection for usage statuses 
- Add get endpoint to get all usage statuses  
- Add validation for usages status when creating and updating an item 

## Testing instructions
- [ ] Review code
- [ ] Check Actions build

## Agile board tracking
closes #230

### CODE REVIEW

File: test/e2e/conftest.py
Comment (joelvdavies) [inline test/e2e/conftest.py:28]:
```suggestion
    Fixture to add usage statuses to the usage_statuses collection in the database before any tests run
```

File: test/e2e/test_item.py
Comment (joelvdavies) [inline test/e2e/test_item.py:971]: I dont think this test is needed `test_partial_update_item` already changes the usage status. Also this is checking the error currently the same as above.

File: README.md
Comment (joelvdavies) [inline README.md:241]: The `Generating new data` should also mention `units and usage statuses`.

File: README.md
Comment (joelvdavies) [inline README.md:206]:
```suggestion
Units and usage statuses should be added to the MongoDB database using `mongoimport` on the provided units and usage statues files found at
```

**Review by joelvdavies — COMMENTED** [review]
A few minor things.  Should be merged with the front end changes.

As a general note this will require us to either write a migration, or delete the existing dev machine data. If we wanted I think this could actually be a nice migration, of course we will have the same for the property ids anyway so there is not much point.

File: README.md
Comment (joelvdavies) [inline README.md:206]:
```suggestion
Units and usage statuses should be added to the MongoDB database using `mongoimport` on the provided units and usage statuses files found at
```

**Comment by joelvdavies** [comment]
@VKTB Should we be attempting to make a migration for this/the property ids? I could take a look if required. I suspect in this particular case its quite simple but the ids are more complicated. If we do I think a script inside the repo itself installed via the `pyproject.toml` would make sense in this particular case as it can use the same credentials and would be needed on the dev machine. Just depends how concerned we are with keeping the existing dev data.

If we do this however this PR should probably wait for that before being merged as it will be slightly easier to test migrations switching between the current dev branch and this.

File: inventory_management_system_api/services/item.py
Comment (VKTB) [inline inventory_management_system_api/services/item.py:86]: Do we need to check this because we are not doing this for units and to me they are both of equal importance? So perhaps we need to do the same check for units if this is actually needed. But the more important question is whether the API should require for `usage_status_id` to be supplied by the client rather than name/value?

File: test/e2e/mock_schemas.py
Comment (VKTB) [inline test/e2e/mock_schemas.py:145]: Do we have to load the usage statuses from the JSON file? Can't we just hardcode a couple of usage statuses (they can even be dummy ones)? I don't think we need to insert all the usage statuses for the purposes of testing anyway.

File: test/e2e/conftest.py
Comment (VKTB) [inline test/e2e/conftest.py:33]: Could this be moved to the `test_usage_status.py` `e2e` test module? Unless I am wrong that's the only module that requires it?

File: test/unit/services/test_item.py
Comment (VKTB) [inline test/unit/services/test_item.py:111]: Unit tests shouldn't interact with the file system as that makes tests less isolated and more susceptible to external factors. Could you please hardcode a couple of usage statuses and use those when mocking instead?

**Comment by VKTB** [comment]
> @VKTB Should we be attempting to make a migration for this/the property ids? I could take a look if required. I suspect in this particular case its quite simple but the ids are more complicated. If we do I think a script inside the repo itself installed via the `pyproject.toml` would make sense in this particular case as it can use the same credentials using pymongo and it would be needed on the dev machine. Just depends how concerned we are with keeping the existing dev data.
> 
> If we do this however, this PR should probably wait for that before being merged as it will be slightly easier to test migrations switching between the current dev branch and this.

We can have a chat about this tomorrow.

File: inventory_management_system_api/services/item.py
Comment (joelvdavies) [reply inventory_management_system_api/services/item.py:86]: We did debate it ourselves, the reason units originally didn't have the check is because we had discussed it as being a placeholder to be edited later, but I do think it could be checked now although I prefer checking using the id rather than pulling the whole list and checking each value, so I feel they are one and the same issue (although we currently have less usage statuses). If we changed this to send back ids, we should do the same for units but as we don't have aggregations and its faster to just store the value, I don't see what the benefit would be to move to ids like that just yet.

File: test/e2e/mock_schemas.py
Comment (joelvdavies) [reply test/e2e/mock_schemas.py:145]: We do the same for units, I agree it doesn't necessarily need all of them but its nice as a check that we haven't broken the syntax of the file. With full crud, this can be tested like the others. We also have the possibility of pre-populating the test data with the units, usage_statuses and predefined catalogue item property using the dev script in the CI if we wanted to (Which is what I have done on the front end api tests), although the expected values would still have to be loaded.

File: test/e2e/conftest.py
Comment (joelvdavies) [reply test/e2e/conftest.py:33]: Its also needed to do the item tests given that the check is in place for the create.

File: inventory_management_system_api/services/item.py
Comment (VKTB) [reply inventory_management_system_api/services/item.py:86]: > the API should require for `usage_status_id` to be supplied by the client rather than name/value?

By this, I meant that the API expects the `usage_status_id` to be supplied by the client in the item `POST` and `PATCH` requests but the API returns the actual usage status name/value (and potentially `usage_status_id`) in the response. This way it would check if the usage status exists using the ID as I also don't like retrieving the whole list and checking each one, and would store both the `usage_status_id` and usage name/value in the database for each item. Storing the IDs means that we won't have to do any ID back population when we get aggregations working nicely in future.

File: test/e2e/mock_schemas.py
Comment (VKTB) [reply test/e2e/mock_schemas.py:145]: I see. Given that we are now doing checks in the API, it means that it now cares what usage status you give it so this functionality now makes sense to stay in the API and with this in mind, I believe that we should support full CRUD (maybe not update and delete to start with as it would require migrations). You mentioned that Joshua wanted to do the full CRUD for the property templates but I see this functionality here as more important. We should do the same for units.

File: test/e2e/conftest.py
Comment (VKTB) [reply test/e2e/conftest.py:33]: Ah yes, sorry, I didn't realise that.

File: test/e2e/mock_schemas.py
Comment (joshuadkitenge) [reply test/e2e/mock_schemas.py:145]: > (maybe not update and delete to start with as it would require migrations)


I'm guessing the patch and delete should work like manufacturers. A user/admin should not be able to delete a usage status if the usage status is being used in an item, and the patch would work if we used IDs instead of the value, but if we use the value, it would be the same as the delete.

Comment (joelvdavies) [reply test/e2e/mock_schemas.py:145]: > I see. Given that we are now doing checks in the API, it means that it now cares what usage status you give it so this functionality now makes sense to stay in the API and with this in mind, I believe that we should support full CRUD (maybe not update and delete to start with as it would require migrations). You mentioned that Joshua wanted to do the full CRUD for the property templates but I see this functionality here as more important. We should do the same for units.

Yeah that sounds sensible to me.



> > (maybe not update and delete to start with as it would require migrations)
> 
> I'm guessing the patch and delete should work like manufacturers. A user/admin should not be able to delete a usage status if the usage status is being used in an item, and the patch would work if we used IDs instead of the value, but if we use the value, it would be the same as the delete.

I think as there is no way to really find those items that use that usage status and it certainly doesn't feel very feasible to expect the user to change them if there are a lot, it does make more sense to me that the edit for an admin functionality like that would require a migration rather than a normal patch as Viktor suggested. Units certainly would pose another challenge in the way that they are stored if we were to try and look up whether they are used or not. Again I think a migration is the only way if they wanted to change the standard unit they use for something for example although I would hope no unit just ceases to exist.

**Comment by VKTB** [comment]
As discussed, please implement the following in this PR:
- `POST` endpoint for creating usage statuses
- `GET` endpoint for getting usage status by ID
- Modify items to take in `usage_status_id`

File: test/unit/services/test_usage_status.py
Comment (joshuadkitenge) [inline test/unit/services/test_usage_status.py:79]:
```suggestion
    # Mock `get` to return a usage status
    retrieved_usage_status = usage_status_service.get(usage_status_id)

    assert retrieved_usage_status is None
    usage_status_repository_mock.get.assert_called_once_with(usage_status_id)
```

File: inventory_management_system_api/services/item.py
Comment (joshuadkitenge) [inline inventory_management_system_api/services/item.py:87]:
```suggestion
            raise MissingRecordError(f"No usage status found with ID: {usage_status_id}")
```

File: inventory_management_system_api/services/item.py
Comment (joelvdavies) [reply inventory_management_system_api/services/item.py]: This is now addressed in https://github.com/ral-facilities/inventory-management-system-api/pull/248

**Review by joelvdavies — CHANGES_REQUESTED** [review]
Mostly minor comments to improve consistency and a couple missing tests. Worked well when I tested it.

File: inventory_management_system_api/repositories/usage_status.py
Comment (VKTB) [inline inventory_management_system_api/repositories/usage_status.py:91]:
```suggestion
        :return: `True` if a duplicate usage status code is found, `False` otherwise
```

File: inventory_management_system_api/repositories/usage_status.py
Comment (VKTB) [inline inventory_management_system_api/repositories/usage_status.py:79]:
```suggestion
        logger.info("Retrieving usage status with ID: %s from the database", usage_status_id)
```

File: inventory_management_system_api/routers/v1/usage_status.py
Comment (joelvdavies) [inline inventory_management_system_api/routers/v1/usage_status.py:39]:
```suggestion
        message = "A usage status with the same value already exists"
```
Just thought this is actually the value.

File: test/unit/repositories/test_usage_status.py
Comment (joelvdavies) [inline test/unit/repositories/test_usage_status.py:24]:
```suggestion
    Test creating a usage status.
```

File: scripts/generate_mock_data.py
Comment (joelvdavies) [inline scripts/generate_mock_data.py:405]:
```suggestion
def create_manufacturer(manufacturer_data: dict) -> dict[str, Any]:
```

File: scripts/generate_mock_data.py
Comment (joelvdavies) [inline scripts/generate_mock_data.py:409]:
```suggestion
def create_usage_status(usage_status_data: dict) -> dict[str, Any]:

```

File: scripts/generate_mock_data.py
Comment (joelvdavies) [inline scripts/generate_mock_data.py:413]:
```suggestion
def create_catalogue_category(category_data: dict) -> dict[str, Any]:

```

File: scripts/generate_mock_data.py
Comment (joelvdavies) [inline scripts/generate_mock_data.py:417]:
```suggestion
def create_catalogue_item(item_data: dict) -> dict[str, Any]:

```

File: scripts/generate_mock_data.py
Comment (joelvdavies) [inline scripts/generate_mock_data.py:421]:
```suggestion
def create_system(system_data: dict) -> dict[str, Any]:

```

File: scripts/generate_mock_data.py
Comment (joelvdavies) [inline scripts/generate_mock_data.py:385]:
```suggestion
def post_avoiding_duplicates(endpoint: str, field: str, json: dict) -> dict[str, Any]:

```

File: scripts/generate_mock_data.py
Comment (joelvdavies) [inline scripts/generate_mock_data.py:425]:
```suggestion
def create_item(item_data: dict) -> dict[str, Any]:
```

**Review by joelvdavies — COMMENTED** [review]
Just two more minor things, I also realised the types were slightly wrong in the generation script.

**Review by VKTB — APPROVED** [review]
Please squash and merge this when ready to be merged.

**Review by joelvdavies — APPROVED** [review]
Do not squash and merge until dev is migrated and functioning properly

File: test/e2e/test_catalogue_item.py
Comment (VKTB) [inline test/e2e/test_catalogue_item.py:2132]:
```suggestion
```

File: test/e2e/test_item.py
Comment (VKTB) [inline test/e2e/test_item.py:109]: Should this be removed?

File: inventory_management_system_api/models/usage_status.py
Comment (VKTB) [inline inventory_management_system_api/models/usage_status.py:16]:
```suggestion
    value: str
```

File: inventory_management_system_api/models/usage_status.py
Comment (VKTB) [inline inventory_management_system_api/models/usage_status.py:18]:
```suggestion
    code: str
```

File: test/e2e/test_item.py
Comment (joshuadkitenge) [reply test/e2e/test_item.py:109]: Yes, but I forgot to remove Line 114

**Review by VKTB — APPROVED** [review]

**Review by joelvdavies — APPROVED** [review]

**Pull request merged**
